worksheets.go

```diff
@@ -30,6 +30,43 @@ type Definitions struct {
 	defs map[string]*Definition
 }
 
+// parentsRefs records and organizes references to all parents of a worksheet,
+// i.e. all worksheets which point directly (ref), or indirectly (e.g. via a
+// ref in a slice) to the worksheet.
+//
+// The map goes from parent's worksheet name, field index of the parent
+// pointing to this worksheet, and then the actual references to the parent
+// worksheet by ID.
+type parentsRefs map[string]map[int]map[string]*Worksheet
+
+func (parents parentsRefs) addParentViaFieldIndex(parent *Worksheet, fieldIndex int) {
+	byParentFieldIndex, ok := parents[parent.def.name]
+	if !ok {
+		byParentFieldIndex = make(map[int]map[string]*Worksheet)
+		parents[parent.def.name] = byParentFieldIndex
+	}
+	byParentId, ok := byParentFieldIndex[fieldIndex]
+	if !ok {
+		byParentId = make(map[string]*Worksheet)
+		byParentFieldIndex[fieldIndex] = byParentId
+	}
+	byParentId[parent.Id()] = parent
+}
+
+func (parents parentsRefs) removeParentViaFieldIndex(parent *Worksheet, fieldIndex int) {
```

Comment:
I find this block hard to reason about because of the repeated identifiers. Would it be OK to do:

`parentName := parent.def.name`

at the top, and the proceed from there?

Reply:
Refactored `removeParentViaFieldIndex`. Can you take another look?

Reply:
Much easier to follow now. Thanks!